### ISSUE TITLE
Questions about full tune the UniAct Model

### ISSUE THREAD

**Issue by Zzfiv3** [2toinf/UniAct#7]
Hi! Thank you for your excellent work and for open-sourcing this project!

I want to full tune the UniAct model based on the Libero_10 dataset. First I follow the [Libero_hdf52jpg.py](https://github.com/2toinf/UniAct/blob/main/datasets/AIRData/Preprocess/Libero_hdf52jpg.py) script and then generate the .pkl file like [Libero.pkl](https://huggingface.co/2toINF/UniAct/blob/main/libero_meta_file.pkl) (**Is there any script to generate the .pkl file?** I wrote a script by myself, but I don't know if it's correct. Let's first assume that my .pkl file is correct and go on). 
Finally I want to evaluate the full tuned UniAct model by using the [run_uniact_libero_eval.py](https://github.com/2toinf/UniAct/blob/main/eval/acthead_libero/run_uniact_libero_eval.py) or [run_uniact_libero_eval.py](https://github.com/2toinf/UniAct/blob/main/eval/libero/run_uniact_libero_eval.py). 

After I run the full tune script 
```
python -m torch.distributed.launch --nproc_per_node=4 \
	slurm_deepspeed_train.py \
        --model UniAct_05B_CodeBook_256_V1_Fulltune \
        --recipe libero_10 \
        --iters 10000 \
        --start_iters 0 \
        --batch-size 16 \
        --lr 1e-4 \
        --grad_accumulation_steps 1 \
        --output_dir exp/Libero_fulltune_MLP \
        --save_interval 10000 \
        --precision bf16 \
        --resume "basemodel.pt"
```
, I only have the full tuned basemodel.pt (named as "mp_rank_00_model_states.pt"), **how could I get the full tuned head params** like [libero_mlp](https://huggingface.co/2toINF/UniAct/blob/main/libero_mlp.pt)?

Thanks!

**Comment by 2toinf** [comment]
Thank you for the interest in UniAct. Given that the entire model has been fine-tuned, mp_rank_00_model_states.pt should contain all parameters, including the head. You can load them at once using load_state_dict and verify success by examining its return message. Feel free to ask if you have further questions.

**Comment by Zzfiv3** [comment]
Thanks for your reply! 

I load the model "mp_rank_00_model_states.pt" and find that the evaluation results are still bad, that could be caused by my wrong way to generate the Libero_10.pkl based on Libero_10 datasets. Could you please provide some help on how to generate Libero_10.pkl file after I run the [Libero_hdf52jpg.py](https://github.com/2toinf/UniAct/blob/main/datasets/AIRData/Preprocess/Libero_hdf52jpg.py) based on Libero_10 datasets?

Also, I may generate wrong data in [config.py](https://github.com/2toinf/UniAct/blob/main/data/AIRData/config.py). After I run [Libero_hdf52jpg.py](https://github.com/2toinf/UniAct/blob/main/datasets/AIRData/Preprocess/Libero_hdf52jpg.py) based on Libero_10, I want to generate two new domains named as "libero_10-1-rgb" and "libero_10-2-rgb". How could I generate the 'action_statics' and 'proprio_statics' in [config.py](https://github.com/2toinf/UniAct/blob/main/data/AIRData/config.py)?

Thanks again!

**Comment by Facebear-ljx** [comment]
Hi!
1. You can change `data_names = ['libero_90']` in Line 7 in [Libero_hdf52jpg.py](https://github.com/2toinf/UniAct/blob/main/data/AIRData/Preprocess/Libero_hdf52jpg.py#L7C1-L7C27) to `data_names = ['libero_10']` to process the libero 10 datasets.

2. Basically, after the hdf52jpg transformation, you can generate the 'action_statics' and 'proprio_statics' by summarizing their min/max/mean/std or the quantile value per-dimension.

3. If you think your generated data are wrong, maybe you can directly use the statistics in [config.py](https://github.com/2toinf/UniAct/blob/main/data/AIRData/config.py) to normalize your proprio and action, since the statistics of Libero-10 are very similar to Libero-130.

**Comment by Zzfiv3** [comment]
Thanks for your reply! 

When I follow the [Fast-Adapt to your embodiment](https://github.com/2toinf/UniAct/tree/main?tab=readme-ov-file#fast-adapt-to-your-embodiment) step, I use [Libero.pkl](https://huggingface.co/2toINF/UniAct/blob/main/libero_meta_file.pkl) and train the model with the following script:
```
python -m torch.distributed.launch --nproc_per_node=4 \
		slurm_deepspeed_train.py \
        --model UniAct_05B_CodeBook_256_V1_For_Fast_Adaptation \
        --recipe ACT-Libero \
        --iters 50000 \
        --start_iters 0 \
        --batch-size 16 \
        --lr 5e-5 \
        --grad_accumulation_steps 1 \
        --output_dir exp/Libero_fastAdapt \
        --save_interval 10000 \
        --precision bf16 \
        --resume "basemodel.pt"
```
with 4 A800 GPUs.  I changed the `--iters` from 1000 to 50000, and the `--lr` from `5e-5` to `1e-4`. But all the evaluation results on Libero are not good. And compared to the provided [MLP head](https://huggingface.co/2toINF/UniAct/blob/main/libero_mlp.pt), the fast-adapt ACT head of my training results are bad. For example, the fast-adapt model is very difficult to finish the tasks on `libero_10`. 

Is there any suggestion on choosing the hyperparameters or any other advices? 

Thanks!

**Comment by 2toinf** [comment]
Hi~ Assuming no other specific errors exist, the primary difference between our setups is the batch size. We utilized a larger batch size, implying that a smaller batch size would necessitate more training iterations. If a large batch size is infeasible, setting hard_forward to False or unfreezing and finetuning the basemodel with a smaller learning rate might accelerate convergence somewhat. I hope this insight is helpful.

**Comment by lyangmam** [comment]
@Zzfiv3 Hi! I’ve encountered a similar issue to the one you mentioned in this thread and I’m trying to resolve it. I’d love to discuss it with you. If it’s convenient, could we continue the conversation via email or another platform? Thanks!

**Comment by Zzfiv3** [comment]
> [@Zzfiv3](https://github.com/Zzfiv3) Hi! I’ve encountered a similar issue to the one you mentioned in this thread and I’m trying to resolve it. I’d love to discuss it with you. If it’s convenient, could we continue the conversation via email or another platform? Thanks!

@lyangmam Of course! We can continue discussing via email~ And it would also be very helpful for me that I can discuss with you!

**Comment by ChangyuLiu2022** [comment]
Hi 

> Thanks for your reply!
> 
> When I follow the [Fast-Adapt to your embodiment](https://github.com/2toinf/UniAct/tree/main?tab=readme-ov-file#fast-adapt-to-your-embodiment) step, I use [Libero.pkl](https://huggingface.co/2toINF/UniAct/blob/main/libero_meta_file.pkl) and train the model with the following script:
> 
> ```
> python -m torch.distributed.launch --nproc_per_node=4 \
> 		slurm_deepspeed_train.py \
>         --model UniAct_05B_CodeBook_256_V1_For_Fast_Adaptation \
>         --recipe ACT-Libero \
>         --iters 50000 \
>         --start_iters 0 \
>         --batch-size 16 \
>         --lr 5e-5 \
>         --grad_accumulation_steps 1 \
>         --output_dir exp/Libero_fastAdapt \
>         --save_interval 10000 \
>         --precision bf16 \
>         --resume "basemodel.pt"
> ```
> 
> with 4 A800 GPUs. I changed the `--iters` from 1000 to 50000, and the `--lr` from `5e-5` to `1e-4`. But all the evaluation results on Libero are not good. And compared to the provided [MLP head](https://huggingface.co/2toINF/UniAct/blob/main/libero_mlp.pt), the fast-adapt ACT head of my training results are bad. For example, the fast-adapt model is very difficult to finish the tasks on `libero_10`.
> 
> Is there any suggestion on choosing the hyperparameters or any other advices?
> 
> Thanks!

Hi～，

When I did [Fast-Adapt to your embodiment] with MLP as the decoder on Libero, I met the same issue. During evaluation, the arm always moved to the center of the table. 

Have you solved this issue? 

Thanks!

**Comment by 2toinf** [comment]
Thank you for your interest in ​UniAct!

We have conducted fast adaptation experiments using an ​ACT head. However, we have not yet trained the ​UniAct_05B_CodeBook_256_V1_For_Fast_Adaptation model with a ​new MLP head, as the original pre-training was performed using an ​MLP head. As a result, we currently lack visibility into how the model behaves after adaptation when using a newly introduced MLP head.

One reason for this caution is that our fast adaptation approach employs the ​​"hard_forward" strategy. While this strategy may perform well in unseen scenarios, it introduces a significant gap between pre-training and post-adaptation performance on in-domain benchmarks. Simply replacing the head with a new ​MLP head does not effectively bridge this gap.

Setting the ​**hard_forward** parameter to ​**False** could help reduce this training gap and improve performance on in-domain tasks.Additionally, appropriate training hyperparameters, such as batch size and total iterations, are always important.

Please let us know if you'd like to discuss this further—we’re happy to provide additional details.

**Comment by ChangyuLiu2022** [comment]
Thank you for your response!

Following your suggestion, I conducted fast adaptation experiments using an ACT head on Libero. I modified the hyperparameters to use --batch_size 256, --iterations 100000, and --learning_rate 5e-5, and ran the training on 4 A6000 GPUs. Unfortunately, the results remain poor, with a success rate of only around 5%.

Could you kindly offer any suggestions or insights on what might be going wrong? Additionally, would you be able to share the hyperparameters you used for fast adaptation with ACT on Libero?

I appreciate your time and help.

**Comment by 2toinf** [comment]
The fine-tuning details are provided in Appendix B.2 of our paper. A success rate of 5 % is highly unusual. As an initial check, try training a smaller ACT model—around 50 million parameters—using your current data pipeline to confirm that your overall setup and data processing are sound. With this configuration, the model should reach a success rate of at least 30–40 %.

**Comment by ChangyuLiu2022** [comment]
Thank you for your response!

The default ACT configuration includes approximately 21 million parameters (well under the 50 million mark), with the majority coming from the Transformer module. The Transformer is initialized in the repository as follows:
self.model = nn.Transformer(
            d_model=hidden_dim,
            nhead=hidden_dim//64,
            num_encoder_layers = 4,
            num_decoder_layers = 2,
            dim_feedforward = hidden_dim * 4,
            dropout = 0.0,
            batch_first = True,
            norm_first = False,
        )

Did I miss anything? I appreciate your patience—thank you!

**Comment by ulzee** [comment]
Hello, thank you for the amazing project. If this discussion was still ongoing I wanted to share what I understood so far and the settings that I have tried.

What I understood is there are two main options (technically three viable options) to train for a robot

0. The various MLP heads that are trained during universal pretraining, although they are not evaluated
1. Full fine tuning (i.e. Fulltune in the code) where the basemodel(?) + control head are both trained.
2. Fast fine tuning - this was done for the real robot and used the ACT head instead of the MLP

Questions: 
* Could you clarify once more how exactly the Libero weights shared in the github were learned? I assume option 1
* but since it is MLP, is this different from the highest scoring method in the paper ("UniAct-0.5B-ACT Head Ours, finetuned")? 
* In fulltune mode is it correct to understand the codebook (essentially the basemodel) is also tuned? Is the basemodel.pt shared in the github tuned for Libero?

With fast fine tuning I was able to obtain a score around >0.7 (=70%?) on libero_spatial within 20k iterations. I wanted to check if my configs look correct as I'm still not sure I'm doing it right:

1. For fast adaptation, the ACT head was recommended and not the MLP head. I set `libero-2-rgb: ACT_2RGB_7DoFs_9Proprio_4FutureAction` in `model_config.py`
2. For fast adaptation with the ACT head, `hard_forward` is recommended to be set to `False`
3. I assume `acthead_libero/run_uniact_libero_eval.py` should be run instead of `libero/run_uniact_libero_eval.py` when evaluating the ACT head
4. I ran `Libero_hdf52jpg` locally but used the Libero.pkl metadata&statistics that was shared by the authors

Questions: 
* What is the distinction between libero-1-rgb and libero-2-rgb?  When does the code decide to use 1 or 2?
* I saw there were several ACT heads defined, which is the correct/best choice?
* How should I interpret the eval scores? In the paper it seemed the best score on spatial was 77%, so can it be expected the fast adaptation can result in `{'sim_summary/libero_spatial/all': 0.74}`?

These were roughly my finetuning cli args:

>     --model UniAct_05B_CodeBook_256_V1_For_Fast_Adaptation \
>     --recipe ACT-Libero \
>     --iters 100000 \
>     --start_iters 0 \
>     --batch-size 32 \
>     --lr 1e-4 \
>     --grad_accumulation_steps 1 \
>     --output_dir ~/data/nvme2/UniAct/exp/Libero \
>     --save_interval 3333 \
>     --port 12345 \
>     --seed 178945 \
>     --precision bf16 \
>     --world_size 1 \
>     --resume downloads/UniAct/basemodel.pt

If there was anything I did not understand correctly I'd appreciate any clarifications! Thanks

**Comment by 2toinf** [comment]
> * Could you clarify once more how exactly the Libero weights shared in the github were learned? I assume option 1

Yes~ The weights we shared for the Libero benchmark were trained during pretraining.

> * but since it is MLP, is this different from the highest scoring method in the paper ("UniAct-0.5B-ACT Head Ours, finetuned")?
The highest-scoring results on Libero were obtained under the Fast-Adaptation setting using the ACT head.

> * In fulltune mode is it correct to understand the codebook (essentially the basemodel) is also tuned? Is the basemodel.pt shared in the github tuned for Libero?

Yes, the codebook is tuned, but the temperature is fixed. All shared parameters in the GitHub repo were obtained during pretraining.

> * What is the distinction between libero-1-rgb and libero-2-rgb?  When does the code decide to use 1 or 2?

It depends on whether you want to include the wrist-view image as an additional observation. Based on our experience, adding wrist-view inputs generally improves performance.

> * I saw there were several ACT heads defined, which is the correct/best choice?

The different ACT head definitions are designed to adapt to varying domains—such as different action/proprioception dimensions or numbers of image views. You can flexibly select one that aligns with the environment requirements.

> * How should I interpret the eval scores? In the paper it seemed the best score on spatial was 77%, so can it be expected the fast adaptation can result in `{'sim_summary/libero_spatial/all': 0.74}`?

If this score is from a single evaluation at a specific checkpoint, then it's entirely reasonable. A 5% fluctuation in success rate is within the expected range. In general, slight hyperparameter adjustments can cause performance variations. The results reported in our paper were carefully tuned and repeatedly tested.
